testsuite/tests/typing-misc/build_as_type.ml

```diff
@@ -0,0 +1,112 @@
+(* TEST
+   * expect
+*)
+
+let f = function ([] : int list) as x -> x ;;
+[%%expect{|
+Line 1, characters 8-42:
+1 | let f = function ([] : int list) as x -> x ;;
+            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Warning 8: this pattern-matching is not exhaustive.
+Here is an example of a case that is not matched:
+_::_
+val f : int list -> int list = <fun>
+|}]
+
+let f =
+  let f' = function ([] : 'a list) as x -> x in
+  f', f';;
+[%%expect{|
+Line 2, characters 11-44:
+2 |   let f' = function ([] : 'a list) as x -> x in
+               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Warning 8: this pattern-matching is not exhaustive.
+Here is an example of a case that is not matched:
+_::_
+val f : ('a list -> 'a list) * ('a list -> 'a list) = (<fun>, <fun>)
```

Comment:
I would naively expect `('a list -> 'a list) * ('b list -> 'b list)` here. I think that the current result is due to the scoping of `'a` to the toplevel binder. Maybe you could point this out in a comment? From a distance the current behavior looks slightly buggy or at least confusing, and I'm not fond of confusing tests without an explanation.

Below there is an example with `([] : _ list)` which gets an even more general type `('a list -> 'b list) * ('c list -> 'd list)`. What should we write if we wanted the intermediate result where `f'` is generalized but the input and output types match? Would `let f' (type a) = function ([] : a list) as x -> x` work correctly?

Reply:
Indeed:
```
# let f (type a) = function ([] : a list) as x -> x in f, f;;
Warning 8: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
_::_
- : ('a list -> 'a list) * ('b list -> 'b list) = (<fun>, <fun>)
```

I'll add it to the testsuite.